File: app/src/main/java/ds/photosight/ui/InfoDialog.java
package ds.photosight.ui;

import java.io.Serializable;
import java.util.Map;

import android.app.Dialog;
import android.content.Context;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.view.*;
import android.view.animation.Animation;
import android.view.animation.Animation.AnimationListener;
import android.view.animation.AnimationUtils;
import android.widget.LinearLayout;
import android.widget.TextView;
import com.androidquery.AQuery;
import com.androidquery.callback.AjaxCallback;
import com.androidquery.callback.AjaxStatus;
import com.androidquery.util.AQUtility;
import ds.photosight.Constants;
import ds.photosight.R;
import ds.photosight.ui.widget.Histogram;
import ds.photosight.utils.L;

public class InfoDialog extends DialogFragment implements Constants {

	Map<Integer, String> data;


	public InfoDialog(Map<Integer, String> map) {
		super();
		data = map;
	}


	public InfoDialog() {
		super();
	}


	@Override
	public void onCreate(Bundle b) {
		setStyle(STYLE_NORMAL, android.R.style.Theme_Panel);

		//getDialog().setOnCancelListener(this);
		if (b != null) {
			data = (Map<Integer, String>) b.get("map");
		}
		super.onCreate(b);
	}


	@Override
	public void onSaveInstanceState(Bundle b) {
		b.putSerializable("map", (Serializable) data);
		super.onSaveInstanceState(b);
	}


	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		WindowManager.LayoutParams lp = getDialog().getWindow().getAttributes();
		lp.dimAmount = 0.4f;
		getDialog().getWindow().setAttributes(lp);
		getDialog().getWindow().addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND);
		final View v = inflater.inflate(R.layout.info, null);

		return v;
	}


	@Override
	public Dialog onCreateDialog(final Bundle savedInstanceState) {
		return new AnimatedDialog(getActivity(), getTheme());
	}


	@Override
	public void onViewCreated(final View v, final Bundle savedInstanceState) {
		super.onViewCreated(v, savedInstanceState);

		final LinearLayout table = (LinearLayout) v.findViewById(R.id.table);
		TextView title = (TextView) v.findViewById(R.id.titleText);
		title.setText(data.get(DATA_IMG_NAME));

		// fetch bitmap
		fetchImage(data);

		table.setVisibility(View.INVISIBLE);
		Animation a = AnimationUtils.loadAnimation(getActivity(), R.anim.anim_in);
		v.startAnimation(a);
		a.setAnimationListener(new AnimationListener() {

			@Override
			public void onAnimationStart(Animation animation) { }


			@Override
			public void onAnimationRepeat(Animation animation) { }


			@Override
			public void onAnimationEnd(Animation animation) {
				table.setVisibility(View.VISIBLE);
				Animation a = AnimationUtils.loadAnimation(getActivity(), R.anim.fade_in);
				table.startAnimation(a);
			}
		});
	}


	private void fetchImage(final Map<Integer, String> data) {

		final AQuery aq = new AQuery(getView());
		Bitmap b = aq.getCachedImage(data.get(DATA_URL_LARGE));
		if (b == null)
			b = aq.getCachedImage(data.get(DATA_URL_SMALL));
		if (b != null)
			((Histogram) aq.id(R.id.histogram).getView()).setBitmap(b);
		else {
			AQUtility.time("fetch");
			aq.ajax(data.get(DATA_URL_LARGE), Bitmap.class, new AjaxCallback<Bitmap>() {
				@Override
				public void callback(final String url, final Bitmap b, final AjaxStatus status) {
					super.callback(url, b, status);
					if (b != null) {
						L.v("bitmap fetched. width=" + b.getWidth());
						AQUtility.timeEnd("fetch", 0);
						((Histogram) aq.id(R.id.histogram).getView()).setBitmap(b);
					}
				}
			});
		}
	}


//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


	private static class AnimatedDialog extends Dialog {

		public AnimatedDialog(final Context context, final int theme) {
			super(context, theme);

		}


		@Override
		public boolean onTouchEvent(final MotionEvent event) {

			if (event.getAction() == MotionEvent.ACTION_DOWN) {
				finalAnimate();
				return true;
			}

			return false;

		}


		@Override
		public void onBackPressed() {
			finalAnimate();
		}


		public void finalAnimate() {

			final View v = findViewById(R.id.infoLayuot);
			final LinearLayout table = (LinearLayout) v.findViewById(R.id.table);

			Animation a = AnimationUtils.loadAnimation(getContext(), R.anim.fade_out);
			table.startAnimation(a);
			a.setAnimationListener(new AnimationListener() {

				@Override
				public void onAnimationStart(Animation animation) {
				}


				@Override
				public void onAnimationRepeat(Animation animation) {
				}


				@Override
				public void onAnimationEnd(Animation animation) {
					table.setVisibility(View.INVISIBLE);
					Animation a = AnimationUtils.loadAnimation(getContext(), R.anim.anim_out);
					v.startAnimation(a);
					a.setAnimationListener(new AnimationListener() {

						@Override
						public void onAnimationStart(Animation animation) { }


						@Override
						public void onAnimationRepeat(Animation animation) { }


						@Override
						public void onAnimationEnd(Animation animation) {
							v.post(new Runnable() {
								@Override
								public void run() {
									dismiss();
								}
							});
						}
					});
				}
			});
		}

	}


}

File: app/src/main/java/ds/photosight/ui/InfoDialog.kt
package ds.photosight.ui

import java.io.Serializable

import android.app.Dialog
import android.content.Context
import android.graphics.Bitmap
import android.os.Bundle
import android.view.*
import android.view.animation.Animation
import android.view.animation.Animation.AnimationListener
import android.view.animation.AnimationUtils
import android.widget.LinearLayout
import android.widget.TextView
import androidx.fragment.app.DialogFragment
import com.androidquery.AQuery
import com.androidquery.callback.AjaxCallback
import com.androidquery.callback.AjaxStatus
import com.androidquery.util.AQUtility
import ds.photosight.Constants
import ds.photosight.R
import ds.photosight.ui.widget.Histogram
import ds.photosight.utils.L

class InfoDialog(var data: Map<Int, String>?) : DialogFragment(), Constants {


    override fun onCreate(b: Bundle?) {
        setStyle(STYLE_NORMAL, android.R.style.Theme_Panel)

        if (b != null) {
            data = b.get("map") as Map<Int, String>
        }
        super.onCreate(b)
    }


    override fun onSaveInstanceState(b: Bundle) {
        b.putSerializable("map", data as Serializable)
        super.onSaveInstanceState(b)
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val lp = dialog!!.window.attributes
        lp.dimAmount = 0.4.toFloat()
        dialog!!.window.attributes = lp
        dialog!!.window.addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
        val v = inflater.inflate(R.layout.info, null)

        return v
    }


    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return AnimatedDialog(activity!!, theme)
    }


    override fun onViewCreated(v: View, savedInstanceState: Bundle?) {
        super.onViewCreated(v, savedInstanceState)

        val table = v.findViewById(R.id.table) as LinearLayout
        val title = v.findViewById(R.id.titleText) as TextView
        title.text = data?.get(Constants.DATA_IMG_NAME)

        // fetch bitmap
        fetchImage(data!!)

        table.visibility = View.INVISIBLE
        val a = AnimationUtils.loadAnimation(activity, R.anim.anim_in)
        v.startAnimation(a)
        a.setAnimationListener(object : AnimationListener {

            override fun onAnimationStart(animation: Animation) {
            }


            override fun onAnimationRepeat(animation: Animation) {
            }


            override fun onAnimationEnd(animation: Animation) {
                table.visibility = View.VISIBLE
                val a = AnimationUtils.loadAnimation(activity, R.anim.fade_in)
                table.startAnimation(a)
            }
        })
    }


    private fun fetchImage(data: Map<Int, String>) {

        val aq = AQuery(view)
        var b: Bitmap? = aq.getCachedImage(data[Constants.DATA_URL_LARGE])
        if (b == null)
            b = aq.getCachedImage(data[Constants.DATA_URL_SMALL])
        if (b != null)
            (aq.id(R.id.histogram).view as Histogram).setBitmap(b)
        else {
            AQUtility.time("fetch")
            aq.ajax(data[Constants.DATA_URL_LARGE], Bitmap::class.java, object : AjaxCallback<Bitmap>() {
                override fun callback(url: String?, b: Bitmap?, status: AjaxStatus?) {
                    super.callback(url, b, status)
                    if (b != null) {
                        L.v("bitmap fetched. width=" + b.width)
                        AQUtility.timeEnd("fetch", 0)
                        (aq.id(R.id.histogram).view as Histogram).setBitmap(b)
                    }
                }
            })
        }
    }


    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


    private class AnimatedDialog(context: Context, theme: Int) : Dialog(context, theme) {


        override fun onTouchEvent(event: MotionEvent): Boolean {

            if (event.action == MotionEvent.ACTION_DOWN) {
                finalAnimate()
                return true
            }

            return false

        }


        override fun onBackPressed() {
            finalAnimate()
        }


        fun finalAnimate() {

            val v: View = findViewById(R.id.infoLayuot)
            val table = v.findViewById(R.id.table) as LinearLayout

            val a = AnimationUtils.loadAnimation(context, R.anim.fade_out)
            table.startAnimation(a)
            a.setAnimationListener(object : AnimationListener {

                override fun onAnimationStart(animation: Animation) {
                }


                override fun onAnimationRepeat(animation: Animation) {
                }


                override fun onAnimationEnd(animation: Animation) {
                    table.visibility = View.INVISIBLE
                    val a = AnimationUtils.loadAnimation(context, R.anim.anim_out)
                    v.startAnimation(a)
                    a.setAnimationListener(object : AnimationListener {

                        override fun onAnimationStart(animation: Animation) {
                        }


                        override fun onAnimationRepeat(animation: Animation) {
                        }


                        override fun onAnimationEnd(animation: Animation) {
                            v.post(object : Runnable {
                                override fun run() {
                                    dismiss()
                                }
                            })
                        }
                    })
                }
            })
        }

    }


}

